### DISCUSSION TITLE
Accessing atoms in one file from imported code

### DISCUSSION THREAD

**Discussion by dbi1** [pmndrs/jotai#1003] [Q&A]
Hi,

I'm using jotai to share state between components. It's a real pleasure to work with, I love the super simple API... Thank you for building jotai!

All works beautifully if all code is in a simple file.

However, things become more complicated if I need to share state between components that are imported from other file(s).

Might anybody have any tips on a good pattern for allowing an imported component to access a jotai atom (or for that matter any variable) in the file that is importing the component?

Here is some simplified code to illustrate what I'd like to do:

```
import React from 'react';
import {atom} from 'jotai';
import ComponentA from './component-a.js';
import ComponentB from './component-b.js';

const myTestAtom = atom({});

export default function App() {
    return (
        <ComponentA></ComponentA>
        <ComponentB></ComponentB>
    );
};
```

Both components would have some internal code that would make use of ```myTestAtom```, like so:

```
const [myTest, setMyTest] = useAtom(myTestAtom);
```

I think my issue is mostly a js scoping issue, but maybe there is a jotai specific pattern that might work well here?

Thank you very much for any thoughts.

**Comment by dai-shi** [comment]
Yes, it's a general issue. Think it like an atom is just like a function (or maybe a variable).

You have ComponentA and ComponentB, and assume both use a common function, which can be say `function add(a, b) { return a + b; }`. How would you organize your code?

There's no specific pattern for atoms. But, yeah, it would be nice if we had some guidelines about structuring atoms and components.

**Comment by dbi1** [comment]
Thank you @dai-shi, that pointed me in a good direction.

I think I have now found a pattern that seems nicely organized and easy to reason about. The idea is to create a separate file (like a store in redux, I think?), where the atom is declared, and which exports the atom.

I then simply need to import the atom in all files where I need it, and use it as if it were declared at the top of that file.

Example file `counterState.js` (where I declare the atom I want to share across files):

```
import { atom } from "jotai";

const myCountAtom = atom(0);

export { myCountAtom };
```

Now I only need to import the atom in all files where I need it, like so:

```
import { useAtom } from "jotai";

import { myCountAtom } from './counterState.js';

MyFunction() {
   const [count, setCount] = useAtom(myCountAtom)
   // ... 
}
```

**Comment by agustind** [comment]
I normally have a "state" folder, like this:


**/state/post.ts:**

```
import { atom } from 'jotai'

export const title = atom('')
export const content = atom('')
```


**/components/PostForm.ts:**

```
import { useAtom } from "jotai";
import * as post from '../state/post'

export default function PostForm(){
    const [title, setTitle] = useAtom(post.title)
    const [content, setContent] = useAtom(post.content)
    
    return (<div>
             <input type="text" value={title} onChange={e => setTitle(e.target.value)} />
             <input type="text" value={content} onChange={e => setContent(e.target.value)} />
       </div>);
}
```



**/components/PostPreview.ts:**

```
import { useAtom } from "jotai";
import * as post from '../state/post'

export default function PostPreview(){
    const [title] = useAtom(post.title)
    const [content] = useAtom(post.content)
    
    return (<div>
        <h1>{title}</h1>
        <p>{content}</p>
       </div>);
}
```
